packages/react-router-redux/modules/ConnectedRouter.js

```diff
@@ -15,22 +15,73 @@ class ConnectedRouter extends Component {
   }
 
   handleLocationChange = location => {
-    this.store.dispatch({
-      type: LOCATION_CHANGE,
-      payload: location
-    })
+    /*
+     * Update the store in 2 cases:
+     * 1) Not syncing history to match store (normal flow from history to store).
+     * 2) DevTools resets state to empty location (need to change to an initial
+     * location).
+    */
+    if (!this.isSyncingHistory || !this.store.getState().router.location.pathname) {
+      this.store.dispatch({
+        type: LOCATION_CHANGE,
+        payload: location
+      })
+    }
+    // Stop syncing the history.
+    this.isSyncingHistory = false
+  }
+
+  handleStoreChange = () => {
+    // Extract store's location.
+    const storeLocation = this.store.getState().router.location
+    const {
+      pathname: storePathname,
+      search: storeSearch,
+      hash: storeHash,
+    } = storeLocation
+    // Extract history's location.
+    const {
+      pathname: historyPathname,
+      search: historySearch,
+      hash: historyHash,
+    } = this.props.history.location
+
+    // In normal situation, the store receives location state from the history.
+    // When the store is updated, the location in the store will be in sync with
+    // the location in the history. However, if the location in the store is
+    // updated before the history, they will be different. We might be time
+    // travelling, replacing root reducer, importing state, etc. So, we need to
+    // update the history to sync with the store.
+    if (historyPathname !== storePathname
+      || historySearch !== storeSearch
+      || historyHash !== storeHash) {
+      // start syncing the history
+      this.isSyncingHistory = true
+
+      if (!storePathname) {
+        // Use initial location if the store is reset.
+        // Store will be updated to initial location in handleLocationChange.
+        this.props.history.push(this.initialLocation)
+      } else {
+        this.props.history.push(storeLocation)
```

Comment:
@mjackson In Hickory, instead of using random keys I use incremental values of `major.minor` where major increments when pushing and minor increments when replacing. Then, determining `n` is just a matter of parsing the keys for the two locations and calculating the difference between majors.

Perhaps I overlooked something when implementing that, but I haven't thought of any reason that that shouldn't work.

I should note that I took the easy route and always use  `window.history` so that I can rely on state, even with hash history. Decided to just support IE10+ and newish Android. http://caniuse.com/#feat=history